diesel/src/upsert/on_conflict_extension.rs

```diff
@@ -238,8 +233,11 @@ impl<T: QuerySource, U, Op, Ret, Target>
     /// [`on_conflict`]: crate::query_builder::InsertStatement::on_conflict()
     pub fn do_nothing(self) -> InsertStatement<T, OnConflictValues<U, Target, DoNothing>, Op, Ret> {
         let target = self.target;
-        self.stmt
-            .replace_values(|values| OnConflictValues::new(values, target, DoNothing))
+        self.stmt.replace_values(|values| {
+            println!("values is {:?}", values);
+            println!("target is {:?}", target);
+            OnConflictValues::new(values, target, DoNothing)
+        })
```

Comment:
Oh whoops didn't mean to commit; was doing this to try and figure out what needed to be done for the `DO NOTHING` case (trying to figure out if we can pass a single column to the downstream AST to do a trivial SET like `ON DUPLICATE KEY UPDATE id=id`

Reply:
Technically we should be able to get the corresponding information from the generic type `T`, which is basically the table. That would require changing the bound above to `T: Table` instead of `T: QuerySource`. That would allow us to get the primary key, but that's not that easy because that key could be a composite key. Adding just a bound there to prevent that is not a great idea as that would prevent the usage of composite primary keys for the other backends as well. It's likely possible to work that around by just moving the type information down to the corresponding `QueryFragment` impl and adding the restrictions there to the mysql specific impl. This likely requires including these information somehow into `DoNothing`.

Docs for `Table`: https://docs.diesel.rs/master/diesel/prelude/trait.Table.html

Reply:
This seems very promising! I am less worried about a primary key being a single column or multiple and more about how to grab the actual name of the columns since ultimately that seems to be [the canonical way to "do nothing"](https://stackoverflow.com/questions/4596390/insert-on-duplicate-key-do-nothing); but since `PrimaryKey` looks to be a trait (bound to other traits) I'm not sure how to actually grab a column name (ultimately would be represented by a string, I would think) 😞

Reply:
> but since PrimaryKey looks to be a trait (bound to other traits) I'm not sure how to actually grab a column name (ultimately would be represented by a string, I would think) 😞

The idea there is not to directly grap the column name, but to forward it to an internal helper. Something like the following could work:

```rust
struct MysqlDoNothingHelper<PK>(PK);

// impl for a single column primary key
impl<C: Column + QueryFragment<Mysql>> QueryFragment<Mysql> for MysqDoNothingHelper<C> {
     fn walk_ast<'b>(&'b self, pass: AstPass<'_, 'b, DB>) -> QueryResult<()> {
        self.0.walk_ast(out.reborrow())?;
        out.push_sql("=");
        self.0.walk_ast(out)?;
        Ok(())
    }
}

// impl for composite primary key impl with two columns
// (checkout `diesel_for_each_tuple!` to automate these impls for all supported tuple sizes
impl<C1: Column + QueryFragment<Mysql>, C2: Column + QueryFragment<Mysql>> QueryFragment<Mysql> for MysqDoNothingHelper<(C1, C2)> {
     fn walk_ast<'b>(&'b self, pass: AstPass<'_, 'b, DB>) -> QueryResult<()> {
        // whatever sql is required for more than one column.
        Ok(())
    }
}
```

Reply:
This is 🤩  but I'm afraid I couldn't quite figure out where I would even use this struct. At least with the code I've got so far it's hard for me to see where I would call this code. Like... in a less strongly typed language, I could do something like `if isinstance(backend, mysql)` I think... but here it looks like you envision having yet another query fragment be built conditionally for only a specific backend (which I think, here, is done via marker traits). But I'm having some trouble wrapping my head around how to do that :(